### ISSUE TITLE
Is Spades limiting threads to 16 in the .info files?

### ISSUE THREAD

**Issue by tseemann** [ablab/spades#112]
We have noticed that despite `--threads 64` being provided, the post-assembly correction bwa-spades is only using 8 threads (see as `bwa-spades -t 8`), and the main spades doesn't seem to use more than 16 threads.

I know that many things don't scale well with more cores (RAM and I/O limited), but I also notice lots of references to `max_threads  16` in the `*.info` setting files.

Is there any chance this is true?

@kwongj

**Comment by asl** [comment]
Hi Torsten,

Could you please provide the spades.log file?

**Comment by tseemann** [comment]
My colleague @kwongj will provide it. I only have these snippets currently:

Called with `-t 64`
```
head -n1 spades.log 
Command line: /home/linuxbrew/.linuxbrew/bin/spades.py    -t    64    --cov-cutoff    auto    --careful    -o    /home/jasonk1/testing/tetyper/Mulcerans/is2404_spades    -1    /home/jasonk1/testing/tetyper/Mulcerans/is2404_mappedreads_1.fq    -2    /home/jasonk1/testing/tetyper/Mulcerans/is2404_mappedreads_2.fq
```

But BWA only using `-t 8`
```
$ cat spades.log |grep "bwa mem"
  0:00:10.369    24M / 24M   INFO   DatasetProcessor         (dataset_processor.cpp     : 149)   Running bwa mem ...:/home/linuxbrew/.linuxbrew/Cellar/spades/3.11.1_2/bin/bwa-spades mem  -v 1 -t 8 /home/jasonk1/testing/tetyper/Mulcerans/is2404_spades/misc/assembled_contigs.fasta /home/jasonk1/testing/tetyper/Mulcerans/is2404_mappedreads_1.fq /home/jasonk1/testing/tetyper/Mulcerans/is2404_mappedreads_2.fq  > /home/jasonk1/testing/tetyper/Mulcerans/is2404_spades/tmp/corrector_04cDl_/lib0_9bd3rW/tmp.sam
  0:00:10.901    24M / 24M   INFO   DatasetProcessor         (dataset_processor.cpp     : 149)   Running bwa mem ...:/home/linuxbrew/.linuxbrew/Cellar/spades/3.11.1_2/bin/bwa-spades mem  -v 1 -t 8 /home/jasonk1/testing/tetyper/Mulcerans/is2404_spades/misc/assembled_scaffolds.fasta /home/jasonk1/testing/tetyper/Mulcerans/is2404_mappedreads_1.fq /home/jasonk1/testing/tetyper/Mulcerans/is2404_mappedreads_2.fq  > /home/jasonk1/testing/tetyper/Mulcerans/is2404_spades/tmp/corrector_I5wQ3Y/lib0_Kb09J5/tmp.sam
```

Is there any chance you are using OpenMP ?
And observing the `$OMP_xxxx` variables?

**Comment by asl** [comment]
Yes, SPAdes is using OpenMP. The variables are being observed by openmp runtime.

**Comment by tseemann** [comment]
I think the mystery is (mostly) solved then - thanks Anton!

Which parts of the pipeline use OpenMP ?
Your version of bwa does?
I don't have this problem with modern BWA MEM.

**Comment by asl** [comment]
> Which parts of the pipeline use OpenMP ?
Everything that is written by us is OpenMP-based :) External tools do use whatever they are written (e.g. bwa mem is a hardcore fork-join pthreads code). SPAdes uses bwa mem 0.7.12

**Comment by mmokrejs** [comment]
Byut spades could probably set properly the environment variable for the downstream tools it executes as its childs. The first is probably not relevant but second is.

```
export NCPUS=64
export OMP_NUM_THREADS=64
```

**Comment by tseemann** [comment]
@asl so if bwa is pthreads, why are you giving it `-t 8` ? when i called spades with `-t 64` ?

**Comment by asl** [comment]
> export OMP_NUM_THREADS=64
BWA does not use OpenMP, so this does not make any sense. All "our" tools honor the global settings and work with OpenMP runtime directly. System / user-level limits should be handled by user themself via setting whatever magic variables the particular environment requires.

**Comment by tseemann** [comment]
@mmokrejs the key variables are
```
OMP_NUM_THREADS=8
OMP_THREAD_LIMIT=64
```
http://gcc.gnu.org/onlinedocs/libgomp/index.html#toc_Environment-Variables

**Comment by asl** [comment]
@tseemann hence I asked for the full log / params.txt file

**Comment by tseemann** [comment]
Ok, i'll wait for @kwongj to respond. We are using default 3.11.0.

**Comment by kwongj** [comment]
[spades.log](https://github.com/ablab/spades/files/1987507/spades.log)

**Comment by kwongj** [comment]
thanks very much for your help @asl !

**Comment by asl** [comment]
Yes, it looks like that OpenMP runtime informs us that maximum 8 threads is allowed. So we honor this limit and use it everywhere (including bwa that is run as a child process for corrector).

PS: It looks like the assembly is very fragmented. Most probably due to very uneven coverage (and assembling in multi-cell mode).

**Comment by kwongj** [comment]
yes, sorry I should have explained that it is not a standard genome assembly - `spades` is being called by a tool that I am currently reviewing the paper for to assemble reads that map to a mobile genetic element, hence the uneven coverage and fragmented assembly!

**Comment by tseemann** [comment]
Can you add this behaviour to the documentation? 
Can't see anything in http://spades.bioinf.spbau.ru/release3.11.0/manual.html

And print something in the log like "Detected OMP_THREAD_LIMIT, limiting all processes to XX threads" ?

**Comment by asl** [comment]
Well, we're not using OMP_THREAD_LIMIT variable directly, OpenMP runtime does. But we will print something like this.
